core/embed/rust/src/trezorhal/storage.rs

```diff
@@ -0,0 +1,330 @@
+use super::ffi;
+use crate::error::Error;
+use core::ptr;
+use cstr_core::{cstr, CStr};
+
+/// Result of PIN delay callback.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PinCallbackResult {
+    /// Continue waiting for PIN unlock.
+    Continue,
+    /// Abort the unlock attempt before time is up.
+    Abort,
+}
+
+/// PIN delay callback function type.
+/// The storage layer will call this function while the PIN timeout is in
+/// progress. This is useful for showing UI progress bar.
+/// `wait` is the total number of seconds waiting.
+/// `progress` is a value between 0 and 1000, where 1000 indicates 100%.
+/// `message` is a message to show to the user.
+pub type PinDelayCallback = fn(wait: u32, progress: u32, message: &str) -> PinCallbackResult;
+
+pub type ExternalSalt = [u8; ffi::EXTERNAL_SALT_SIZE as usize];
+
+/// Static reference to the currently set PIN callback function.
+static mut PIN_UI_CALLBACK: Option<PinDelayCallback> = None;
+
+/// C-compatible wrapper for the Rust callback.
+unsafe extern "C" fn callback_wrapper(
+    wait: u32,
+    progress: u32,
+    message: *const cty::c_char,
+) -> ffi::secbool {
+    let message = unsafe { CStr::from_ptr(message as _) };
+    let result = unsafe {
+        PIN_UI_CALLBACK
+            .map(|c| c(wait, progress, message.to_str().unwrap_or("")))
+            .unwrap_or(PinCallbackResult::Continue)
+    };
+    if matches!(result, PinCallbackResult::Abort) {
+        ffi::sectrue
+    } else {
+        ffi::secfalse
+    }
+}
+
+/// Storage error type.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum StorageError {
+    /// Data is too long or not appropriate for the operation.
+    InvalidData,
+    /// Write failed.
+    WriteFailed,
+    /// Read failed.
+    ReadFailed,
+    /// Delete failed.
+    DeleteFailed,
+    /// Failed to perform the get-and-increment operation on a counter.
+    CounterFailed,
+}
+
+impl From<StorageError> for Error {
+    fn from(err: StorageError) -> Self {
+        match err {
+            StorageError::InvalidData => Error::ValueError(cstr!("Invalid data for storage")),
+            StorageError::WriteFailed => Error::ValueError(cstr!("Storage write failed")),
+            StorageError::ReadFailed => Error::ValueError(cstr!("Storage read failed")),
+            StorageError::DeleteFailed => Error::ValueError(cstr!("Storage delete failed")),
+            StorageError::CounterFailed => {
+                Error::ValueError(cstr!("Retrieving counter value failed"))
+            }
+        }
+    }
+}
+
+pub type StorageResult<T> = Result<T, StorageError>;
+
+/// Initialize the storage layer.
+/// This function must be called before any other storage function.
+pub fn init() {
+    unsafe {
+        ffi::storage_init(
+            Some(callback_wrapper),
```

Comment:
Seems to be crashing in this function. This line looks suspicious to me.

Reply:
Nevermind, the `Some` is actually fine.

Reply:
Maybe we need to call `flash_init` in the test setup?

Reply:
it's possible that it became necessary in the meantime .)

Reply:
yeah that works! thanks https://github.com/trezor/trezor-firmware/commit/e6770444d6e4b67c5fcd7244030bb0f103e02639